feat: add web driver setup details for flow-client (#1257)
vaadin/docs: articles/contributing/flow-unit-testing.asciidoc

layout: page
---

= Unit Testing in Flow
:experimental:
:commandkey: &#8984;

Every change in the code base requires a JUnit test for the code change.

> In cases where a JUnit test is not plausible then an <<flow-integration-testing#,Integration Test>> should be added instead.

Reviewer: ```suggestion
In cases where a JUnit test is not practicable, an <<flow-integration-testing#,Integration Test>> should be added instead.
```

== How to Run the Tests

Running the following command results in running all the tests in the specified module:

[source,terminal]
----

> mvn verify -pl <module-name>

Reviewer 2: This should probably mention that it works like this only for direct children of the directory as for instance running
`mvn test -pl flow-maven-plugin` in root will result in
```
[INFO] Scanning for projects...
[ERROR] [ERROR] Could not find the selected project in the reactor: flow-maven-plugin @ 
[ERROR] Could not find the selected project in the reactor: flow-maven-plugin -> [Help 1]
```
Where as running `mvn test -pl :flow-maven-plugin` will run the maven plugin module which is located at `./flow-plugins/flow-maven-plugin` optionally one could have the full path `mvn test -pl flow-plugins/flow-maven-plugin`

Author: Added a TIP to include details for running tests in nested modules.

----

You can execute tests for single class with the following command:

[source,terminal]
----
mvn -Dtest=<test-class-name> test -pl <module-folder-name>
----

> Also, for running a single test inside a class:

Reviewer: ```suggestion
Also, to run a single test inside a class:
```

[source,terminal]
----
mvn -Dtest=<test-class-name>#<test-method-name> test -pl <module-folder-name>
----

> For running all the unit tests in the project:

Reviewer: ```suggestion
To run all the unit tests in the project:
```

[source,terminal]
----
mvn test -am -pl flow
----

> For running tests through your IDE, please refer to the documentation of your desired IDE.

Reviewer: ```suggestion
To run tests via your IDE, please refer to the IDE's documentation.
```

> .Chrome Web Driver is required

Reviewer 2: If this is true then we should add the local-run profile to the client module pom to have it install the drivers automatically according to the `${project.rootdir}/drivers.xml` file
